x/precisebank/keeper/send.go

```diff
@@ -19,23 +29,188 @@ func (k Keeper) SendCoins(
 	// IsSendEnabledCoins() is only used in x/bank in msg server, not in keeper,
 	// so we should also not use it here to align with x/bank behavior.
 
-	panic("unimplemented")
+	if !amt.IsValid() {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidCoins, amt.String())
+	}
+
+	passthroughCoins := amt
+	extendedCoinAmount := amt.AmountOf(types.ExtendedCoinDenom)
+
+	// Remove the extended coin amount from the passthrough coins
+	if extendedCoinAmount.IsPositive() {
+		subCoin := sdk.NewCoin(types.ExtendedCoinDenom, extendedCoinAmount)
+		passthroughCoins = amt.Sub(subCoin)
+	}
+
+	// Send the passthrough coins through x/bank
+	if passthroughCoins.IsAllPositive() {
+		if err := k.bk.SendCoins(ctx, from, to, passthroughCoins); err != nil {
+			return err
+		}
+	}
+
+	// If there is no extended coin amount, we are done
+	if extendedCoinAmount.IsZero() {
+		return nil
+	}
+
+	// Send the extended coin amount through x/precisebank
+	return k.sendExtendedCoins(ctx, from, to, extendedCoinAmount)
+}
+
+// sendExtendedCoins transfers amt extended coins from a sending account to a
+// receiving account. An error is returned upon failure. This function is
+// called by SendCoins() and should not be called directly.
+func (k Keeper) sendExtendedCoins(
+	ctx sdk.Context,
+	from, to sdk.AccAddress,
+	amt sdkmath.Int,
+) error {
+	// Check balance is sufficient
+	bal := k.GetBalance(ctx, from, types.ExtendedCoinDenom)
+	if bal.Amount.LT(amt) {
+		coin := sdk.NewCoin(types.ExtendedCoinDenom, amt)
+
+		// TODO: This checks spendable coins and returns error with spendable
+		// coins, not full balance. If GetBalance() is modified to return the
+		// full, including locked, balance then this should be updated to deduct
+		// locked coins.
+
+		// Use sdk.NewCoins() so that it removes empty balances - ie. prints
+		// empty string if balance is 0. This is to match x/bank behavior.
+		spendable := sdk.NewCoins(bal)
+
+		return errorsmod.Wrapf(
+			sdkerrors.ErrInsufficientFunds,
+			"spendable balance %s is smaller than %s",
+			spendable, coin,
+		)
+	}
+
+	integerAmt := amt.Quo(types.ConversionFactor())
+	fractionalAmt := amt.Mod(types.ConversionFactor())
+
+	// Account old balances
+	senderFractionalBal := k.GetFractionalBalance(ctx, from)
+	recipientFractionalBal := k.GetFractionalBalance(ctx, to)
+
+	// Account new fractional balances (NOT YET carried)
+	senderNewFractionalBal := senderFractionalBal.Sub(fractionalAmt)
+	recipientNewFractionalBal := recipientFractionalBal.Add(fractionalAmt)
+
+	// Check if sender needs to borrow and recipient needs to carry
+	senderNeedsBorrow := senderNewFractionalBal.IsNegative()
+	recipientNeedsCarry := recipientNewFractionalBal.GTE(types.ConversionFactor())
+
```

Comment:
i'm struck by how complex the control flow is in this method after this point. i'm trying to wrap my head around the tradeoff being made.

the added checks allow us to prevent two unnecessary transfers (1 ukava from sender to module, then 1 ukava from module to recipient) by allowing the sender to cover the recipient's borrow if it can be covered by their carry: they just send one more integer unit instead of passing the ukava through this module's account.

i wonder about if we could reduce complexity by just always handling the carryover & borrow in the same way, regardless of the fact that 1 out of the 4 possible cases could avoid the two transfers to/from precisebank.

possible cases:
1. sender's balance >= sent, receiver's new balance < conversion factor
2. sender's balance < sent, receiver's new balance < conversion factor
3. sender's balance >= sent, receiver's new balance > conversion factor
4. sender's balance < sent, receiver's new balance > conversion factor

only in case 4 is it possible to cover the receiver's necessary borrow with the sender's carryover; otherwise, some part of the sender -> precisebank -> receiver ukava transfers must occur.

i'm not sure what to make of it. i think my point is that this method just requires a lot of mental overhead to think through, and it feels like there is a refactor or tradeoff asking to be made for the sake of comprehension and readability.

i think the code is correct (and properly tested in all cases!), just wondering if we can spend some time trying to formulate it in an easier to grok way. maybe separating each of the above cases and handling them, with some non-DRY code? or foregoing the optimization in case 4?

i am still thinking it through.... i think the mental overhead is coming from the separated `canDirectTransferCarry` / `!canDirectTransferCarry` conditions.

Reply:
this is what i arrived at. i think that making explicit the 4 possible cases in the code allows for a clearer understanding of what's happening here, all while still keeping onlly the minimum number of necessary transfers.

```go
senderNeedsBorrow := senderNewFractionalBal.IsNegative()
recipientNeedsCarry := recipientNewFractionalBal.GTE(types.ConversionFactor())

// handle adjusting balances necessary for fractional balance carries/borrows
switch {
// 1. sender's balance >= sent, receiver's new balance < conversion factor
case !senderNeedsBorrow && !recipientNeedsCarry:
	{
		// happy path! no updates necessary
	}
// 2. sender's balance < sent, receiver's new balance < conversion factor
case senderNeedsBorrow && !recipientNeedsCarry:
	{
		borrowCoin := sdk.NewCoin(types.IntegerCoinDenom, sdk.NewInt(1))
		if err := k.bk.SendCoinsFromAccountToModule(ctx, from, types.ModuleName, sdk.NewCoins(borrowCoin)); err != nil {
			return err
		}
		senderNewFractionalBal = senderNewFractionalBal.Add(types.ConversionFactor())
	}
// 3. sender's balance >= sent, receiver's new balance > conversion factor
case !senderNeedsBorrow && recipientNeedsCarry:
	{
		carryCoin := sdk.NewCoin(types.IntegerCoinDenom, sdk.NewInt(1))
		if err := k.bk.SendCoinsFromModuleToAccount(ctx, types.ModuleName, to, sdk.NewCoins(carryCoin)); err != nil {
			return err
		}
		recipientNewFractionalBal = recipientNewFractionalBal.Sub(types.ConversionFactor())
	}
	// 4. sender's balance < sent, receiver's new balance > conversion factor
case senderNeedsBorrow && recipientNeedsCarry:
	{
		// direct transfer possible
		// add 1 to integer amount
		integerAmt = integerAmt.AddRaw(1)
		senderNewFractionalBal = senderNewFractionalBal.Add(types.ConversionFactor())
		recipientNewFractionalBal = recipientNewFractionalBal.Sub(types.ConversionFactor())
	}
}

if integerAmt.IsPositive() {
	transferCoin := sdk.NewCoin(types.IntegerCoinDenom, integerAmt)
	if err := k.bk.SendCoins(ctx, from, to, sdk.NewCoins(transferCoin)); err != nil {
		return err
	}
}
```

i'm not tied to this, but i think it's worth trying to make this (very important!) logic easier to follow. but also be sure to add back in all your great comments 😆

Reply:
Valid points @pirtleshell, but I'm not sure the case statement helps here.  I think @drklee3 portrays the logic well, but we can make a couple simplifications by using mod instead of Add/Sub ConversionFactor, and remove the named direct transfer variable to flatten the logic.  I also removed the up front balance check to reduce gas costs for successful transfers.

Thoughts @drklee3 @pirtleshell ?

```golang
// sendExtendedCoins transfers amt extended coins from a sending account to a
// receiving account. An error is returned upon failure. This function is
// called by SendCoins() and should not be called directly.
func (k Keeper) sendExtendedCoins(
	ctx sdk.Context,
	from, to sdk.AccAddress,
	amt sdkmath.Int,
) error {
	integerSendAmt := amt.Quo(types.ConversionFactor())

	senderFractionalBal := k.GetFractionalBalance(ctx, from)
	senderNewFractionalBal := senderFractionalBal.Sub(amt).Mod(types.ConversionFactor())

	recipientFractionalBal := k.GetFractionalBalance(ctx, to)
	recipientNewFractionalBal := recipientFractionalBal.Add(amt).Mod(types.ConversionFactor())

	// we went past zero and wrapped around, must decrease sender balance by 1
	borrowRequired := senderNewFractionalBal.GT(senderFractionalBal)
	// we went past ConversionFactor-1 and wrapped around, must increase receipient balance by 1
	carryRequired := recipientNewFractionalBal.LT(recipientFractionalBal)

	// we can skip the reserves and increase transfer amount by 1
	if carryRequired && borrowRequired {
		integerSendAmt = integerSendAmt.Add(sdkmath.OneInt())
	}

	// transfer integer amount from sender to receipient
	// we do the the larger transfer first to fail fast
	if integerSendAmt.IsPositive() {
		transferCoin := sdk.NewCoin(types.IntegerCoinDenom, integerSendAmt)
		if err := k.bk.SendCoins(ctx, from, to, sdk.NewCoins(transferCoin)); err != nil {
			// We assume most transactions will succeed, so skip checking the balance ahead of time
			// to save gas cost.  The bank module will check spendable balance for us.
			return k.wrapError(ctx, from, amt, err)
		}
	}

	// we didn't borrow from sender during transfer, so take 1 from sender and increase reserves
	if borrowRequired && !carryRequired {
		borrowCoin := sdk.NewCoin(types.IntegerCoinDenom, sdkmath.OneInt())
		if err := k.bk.SendCoinsFromAccountToModule(ctx, from, types.ModuleName, sdk.NewCoins(borrowCoin)); err != nil {
			// We assume most transactions will succeed, so skip checking the balance ahead of time
			// to save gas cost.  The bank module will check spendable balance for us.
			return k.wrapError(ctx, from, amt, err)
		}
	}

	// we didn't carry to recipient during transfer, so take 1 from reserve and decrease reserve
	if carryRequired && !borrowRequired {
		carryCoin := sdk.NewCoin(types.IntegerCoinDenom, sdkmath.OneInt())
		if err := k.bk.SendCoinsFromModuleToAccount(ctx, types.ModuleName, to, sdk.NewCoins(carryCoin)); err != nil {
			// This will only error if state is invalid, we should panic as caller can not fix
			panic(err)
		}
	}

	// update fractional balances
	k.SetFractionalBalance(ctx, from, senderNewFractionalBal)
	k.SetFractionalBalance(ctx, to, recipientNewFractionalBal)

	return nil
}

func (k Keeper) wrapError(ctx sdk.Context, addr sdk.AccAddress, amt sdk.Int, err error) error {
	if !errors.Is(err, sdkerrors.ErrInsufficientFunds) {
		return err
	}

	// Check balance is sufficient
	bal := k.GetBalance(ctx, addr, types.ExtendedCoinDenom)
	coin := sdk.NewCoin(types.ExtendedCoinDenom, amt)

	// TODO: This checks spendable coins and returns error with spendable
	// coins, not full balance. If GetBalance() is modified to return the
	// full, including locked, balance then this should be updated to deduct
	// locked coins.

	// Use sdk.NewCoins() so that it removes empty balances - ie. prints
	// empty string if balance is 0. This is to match x/bank behavior.
	spendable := sdk.NewCoins(bal)

	return errorsmod.Wrapf(
		sdkerrors.ErrInsufficientFunds,
		"spendable balance %s is smaller than %s",
		spendable, coin,
	)
}
```

Reply:
nice, @nddeluca 👍 

that is closer to what i had before i pulled the carry & borrow transfers up into the switch-case. i prefer your separation of pure and statefulness. particularly i think this is great for comprehension:
```go
transferBorrowToReserve := borrowRequired && !carryRequired
transferCarryToRecipient := carryRequired && !borrowRequired
if borrowRequired && carryRequired {
	integerSendAmt = integerSendAmt.Add(sdkmath.OneInt())
}
```
also some great pure function extractions ✨ 

---

when i was thinking through this i was under the impression we needed to check for balance at the start but i'm trying to remember why.. but, if they don't have sufficient balance one of these transfers will fail:
1. integer transfer of ukava to the receiver
2. borrow transfer of 1 ukava to the module account

seems correct to optimize gas costs for successful txs by relying on those to error if there's insufficient balance 👍

Reply:
all great points 😃 just confirmed that even if its only a fractional send and fractional balance is not sufficient (e.g 0.02, send 0.03) that still incurs a borrow and thus the error - so a preemptive balance check is not necessary. Will also include those 4 cases in a doc comment since that's also useful to understand